tower-http/src/services/fs/serve_dir.rs

```diff
@@ -406,58 +456,118 @@ impl Future for ResponseFuture {
+                                        .boxed();
+                                ResponseBody::new(box_body)
+                            } else {
+                                empty_body()
+                            };
+                            builder.body(body)
+                        }
                     }
 
-                    Poll::Ready(Ok(res))
-                }
-
-                Ok(Output::FileMetadata(metadata, mime, maybe_encoding)) => {
-                    let mut res = Response::new(empty_body());
-                    res.headers_mut().insert(header::CONTENT_TYPE, mime);
-                    res.headers_mut()
-                        .insert(header::CONTENT_LENGTH, metadata.len().into());
-
-                    if let Some(encoding) = maybe_encoding {
-                        res.headers_mut()
-                            .insert(header::CONTENT_ENCODING, encoding.into_header_value());
+                    Ok(Output::Redirect(location)) => {
+                        let res = Response::builder()
+                            .header(http::header::LOCATION, location)
+                            .status(StatusCode::TEMPORARY_REDIRECT)
+                            .body(empty_body())
+                            .unwrap();
+                        return Poll::Ready(Ok(res));
                     }
 
-                    Poll::Ready(Ok(res))
-                }
-
-                Ok(Output::Redirect(location)) => {
-                    let res = Response::builder()
-                        .header(http::header::LOCATION, location)
-                        .status(StatusCode::TEMPORARY_REDIRECT)
-                        .body(empty_body())
-                        .unwrap();
-                    Poll::Ready(Ok(res))
-                }
-
-                Ok(Output::NotFound) => {
-                    let res = Response::builder()
-                        .status(StatusCode::NOT_FOUND)
-                        .body(empty_body())
-                        .unwrap();
+                    Ok(Output::NotFound) => {
+                        let res = Response::builder()
+                            .status(StatusCode::NOT_FOUND)
+                            .body(empty_body())
+                            .unwrap();
 
-                    Poll::Ready(Ok(res))
-                }
+                        return Poll::Ready(Ok(res));
+                    }
 
-                Err(err) => Poll::Ready(
-                    super::response_from_io_error(err).map(|res| res.map(ResponseBody::new)),
-                ),
-            },
+                    Err(err) => {
+                        return Poll::Ready(
+                            super::response_from_io_error(err)
+                                .map(|res| res.map(ResponseBody::new)),
+                        )
+                    }
+                };
+                Poll::Ready(Ok(response.unwrap()))
```

Comment:
Is it worth returning the match directly to avoid having to have this Poll::Ready so far down? Maybe not a problem if the range stuff gets broken out to separate function